### PR TITLE
JAMES-3184 Make throttling more resilient

### PR DESCRIPTION
**Pull request by chibenwa**
- Large flux handling
 - Error handling (don't crash when there is an error)

### CODE REVIEW

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 1]
```diff
@@ -151,6 +152,120 @@ void throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize() {
...
+                .collectList()
+                .block();
+
+            assertThat(results)
+                .containsExactly(0, 1, 2, 3, 4, 6, 7, 8, 9);
+        }
+
+        @Test
+        void throttleShouldNotAbortProcessingUponUpstreamError() {
+            Hooks.onOperatorDebug();
```
Comment (LanKhuat) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:184]: do we need this?

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 2]
```diff
@@ -151,6 +152,120 @@ void throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize() {
...
+            Duration windowDuration = Duration.ofMillis(100);
+
+            Flux<Integer> originalFlux = Flux.range(0, 10);
+            Function<Integer, Publisher<Integer>> longRunningOperation =
+                i -> {
+                    if (i == 5) {
+                        return Mono.error(new RuntimeException());
+                    }
+                    return Mono.just(i);
+                };
```
Comment (LanKhuat) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:168]: Not that long of an operation :) ?

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 3]
```diff
@@ -151,6 +152,120 @@ void throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize() {
...
+
+            Flux<Long> originalFlux = Flux.interval(Duration.ofMillis(10));
+
+            ImmutableList<Long> perWindowCount = originalFlux
+                .transform(ReactorUtils.<Long, Long>throttle()
+                    .elements(windowMaxSize)
+                    .per(windowDuration)
+                    .forOperation(i -> Mono.fromCallable(() -> stopwatch.elapsed(TimeUnit.MILLISECONDS))))
+                .map(i -> i / 20)
+                .doOnSubscribe(signal -> stopwatch.start())
```
Comment (LanKhuat) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:238]: what is the usage of stopwatch if we don't assert on it?

File: server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java [hunk 4]
```diff
@@ -54,7 +54,7 @@
             Preconditions.checkArgument(!duration.isZero(), "'windowDuration' must be strictly positive");
 
             return flux -> flux
-                .windowTimeout(elements, duration)
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:56]: the use case not covered AFAIU is when you don't have a multiple of `elements`, your window never fill and the pipeline never continue to the `zipWith`. Could you write a test with less elements in the Flux than the `elements` parameter to check that case?

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java (hunk 3 above)
Comment (chibenwa) [reply server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:238]: stopWatch is here to tell "which window" an element belongs to.

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java (hunk 2 above)
Comment (chibenwa) [reply server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:168]: Let's rename it operation

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java (hunk 1 above)
Comment (chibenwa) [reply server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:184]: No

File: server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java (hunk 4 above)
Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:56]: I think the termination signal would do the job, as long as we don't throttle infinite streams.

File: server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java [hunk 5]
```diff
@@ -53,10 +54,11 @@
...
+                .window(elements)
                 .zipWith(Flux.interval(DELAY, duration))
                 .flatMap(Tuple2::getT1, elements, elements)
-                .flatMap(operation, elements);
+                .flatMap(operation, elements)
+                .onErrorContinue((e, o) -> LOGGER.error("Error encountered while throttling for {}", o.toString(), e));
```
Comment (LanKhuat) [inline server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:61]: toString() is causing NullPointerException on james-charge-01

Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:61]: Good reason to make it implicit then?

**Comment by chibenwa** [comment]
Another issue is "can we cancel throttling" ? We need further investigations...

File: server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java (hunk 4 above)
Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:56]: Short version: thanks for this feedback, it lead to write more tests, detect more issues and hopefully upgrade the code.

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 6]
```diff
@@ -264,6 +268,117 @@ void throttleShouldNotDropEntriesWhenUpstreamIsSlow() {
 
             assertThat(results).containsExactly(0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L);
         }
+
+        @Test
+        void throttleShouldCompleteWhenOriginalFluxDoNotFillAWindow() {
```
Comment (Arsnael) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:277]:
```suggestion
        void throttleShouldCompleteWhenOriginalFluxDoesNotFillAWindow() {
```

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 7]
```diff
@@ -264,6 +268,117 @@ void throttleShouldNotDropEntriesWhenUpstreamIsSlow() {
...
+                .collectList()
+                .block();
+
+            System.out.println(results);
+            assertThat(results).containsExactly(0L, 1L, 2L);
+        }
+
+        @Test
+        void throttleShouldTolerateManyEmptySuccessiveWindows() {
+            Hooks.onOperatorDebug();
```
Comment (Arsnael) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:342]: not necessary?

**Comment by chibenwa** [comment]
```

[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537] [ERROR]   CassandraReIndexerImplTest$FailureTesting.fullReindexingShouldUpdateDetailsUponSingleMessageFullReadError:327 
[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537] Expecting:
[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537]   <[]>
[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537] to contain exactly (and in same order):
[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537]   <[ReIndexingFailure{mailboxId=CassandraId{id=088b76b0-bac6-11ea-871d-218c014144ac}, uid=MessageUid{uid=1}}]>
[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537] but could not find the following elements:
[9d2c16a58eb32ee45a2f24fa533e4e3d4baf8537]   <[ReIndexingFailure{mailboxId=CassandraId{id=088b76b0-bac6-11ea-871d-218c014144ac}, uid=MessageUid{uid=1}}]>
```

Investigating

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
`RabbitMQDeletedMessageVaultIntegrationTest>DeletedMessageVaultIntegrationTest.vaultEndpointShouldRestoreImapDeletedMailbox:255 » ConditionTimeout`

unrelated

test this please

File: server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java [hunk 8]
```diff
@@ -38,25 +40,26 @@
...
 
             return flux -> flux
-                .windowTimeout(elements, duration)
-                .zipWith(Flux.interval(DELAY, duration))
-                .flatMap(Tuple2::getT1, elements, elements)
-                .flatMap(operation, elements);
+                .onErrorContinue((e, o) -> LOGGER.error("Error encountered while generating throttled entries", e))
+                .window(elements)
+                .delayElements(duration)
+                .concatMap(window -> window.flatMap(operation)
```
Comment (blackheaven) [inline server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:58]: I do not get how it work, can you explain it?

Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java:58]: We produce fixed windows, each one get delayed, and for each one we apply the transformation.

From the previous version, it resolves some reactor bugs. See tests for limitations.

**Comment by chibenwa** [comment]
>" You need to leave a comment indicating the requested changes. "

Sorry, but unless someone is clever enough to come up with an alternative, this is our best chance to hold the indexing time engagement.

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 9]
```diff
@@ -151,6 +156,247 @@ void throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize() {
...
+            Duration windowDuration = Duration.ofMillis(1);
+
+            Flux<Integer> originalFlux = Flux.range(0, 10000);
+
+            originalFlux
+                .transform(ReactorUtils.<Integer, Integer>throttle()
+                    .elements(windowMaxSize)
+                    .per(windowDuration)
+                    .forOperation(Mono::just))
+                .blockLast();
```
Comment (rouazana) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java:242]: no assertion?

**Comment by chibenwa** [comment]
Merged.

**Pull request closed without merging**